content/blog/2022/2022-01-25-annotation-processing.md

```diff
@@ -0,0 +1,951 @@
+
+        System.out.println("Employee Id: " + id);
+        System.out.println("Employee Name: " + name);
+    }
+}
+```
+
+Run a test as below:
+
+```
+public class TestSingleValueAnnotatedEmployee {
+
+    public static void main(String[] args) {
+
+        SingleValueAnnotatedEmployee employee = new SingleValueAnnotatedEmployee(1, "John Doe");
+        employee.getEmployeeDetails();
+
+        Annotation companyAnnotation = employee.getClass().getAnnotation(SingleValueAnnotationCompany.class);
+        SingleValueAnnotationCompany company = (SingleValueAnnotationCompany)companyAnnotation;
+
+        System.out.println("Company Name: " + company.value());
+    }
+}
+```
+
+The single value 'XYZ' overrides the default annotation value and the output is as below:
+
+```
+Employee Id: 1
+Employee Name: John Doe
+Company Name: XYZ
+```
+
+### Full Annotations
+
+They consist of multiple name value pairs. For eg. Company(name="ABC", city="XYZ"). Considering our original Company example:
+
+```java
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Company{
+    String name() default "ABC";
+    String city() default "XYZ";
+}
+```
+
+Create the MultiValueAnnotatedEmployee class as below. Specify the parameters and values as below. The default values will be over-written.
+
+```java
+@Company(name = "AAA", city = "ZZZ")
+public class MultiValueAnnotatedEmployee {
+    
+}
+```
+
+Run a test as below:
+
+```java
+public class TestMultiValueAnnotatedEmployee {
+
+    public static void main(String[] args) {
+
+        MultiValueAnnotatedEmployee employee = new MultiValueAnnotatedEmployee();
+
+        Annotation companyAnnotation = employee.getClass().getAnnotation(Company.class);
+        Company company = (Company)companyAnnotation;
+
+        System.out.println("Company Name: " + company.name());
+        System.out.println("Company City: " + company.city());
+    }
+}
+```
+
+The output is as below, and has overridden the default annotation values:
+
+```
+Company Name: AAA
+Company City: ZZZ
+```
+
+
+
+## A real-world example
+
+For our real-world example, we are going to do a simple simulation of the annotation @Test in JUnit. By marking our functions with the @Test annotation we can determine at run-time, the methods in a class that need to be run as tests. 
+
+We first create the Annotation as below:
+
+```java
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD) 
+public @interface Test {
+}
+```
+
+Next, we create a class AnnotatedMethods, to which we will apply the @Test annotations to some of its methods:
+
+```java
+public class AnnotatedMethods {
+
+    @Test
+    public void Test_1() {
+
+        System.out.println("This is the first test");
+    }
+
+    public void Test_2() {
+
+        System.out.println("This is the second test");
+    }
+}
+
+```
+
+Now we create the test to run AnnotatedMethods as below. The main steps are:
+
+1. Get a reference to the class:
```

Comment:
Explaining a code example line-by-line like this is very hard to read and understand. Since the code example is not terribly long, I would just write a sentence to introduce the code example, show the code example, and then explain the steps referencing the code example above in the text (without repeating the code verbatim). One paragraph per step. You don't have to be too specific when explaining the code, because it's part of the learning experience to connect the dots.

I.e. something like this:

By calling `getDeclaredMethods()`, we're getting the methods of our `AnnotatedMethods` class.

Then, we're iterating through the methods and check for each method if it is annotated with our `@Test` annotation.

...